pkg/controller/stacks/clusterrequest/clusterrequest.go

```diff
@@ -0,0 +1,567 @@
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package clusterrequest
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"io"
+	"os"
+	"strings"
+	"sync"
+	"time"
+
+	batchv1 "k8s.io/api/batch/v1"
+	corev1 "k8s.io/api/core/v1"
+	kerrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/yaml"
+	"k8s.io/client-go/kubernetes"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	"github.com/pkg/errors"
+
+	runtimev1alpha1 "github.com/crossplaneio/crossplane-runtime/apis/core/v1alpha1"
+	"github.com/crossplaneio/crossplane-runtime/pkg/logging"
+	"github.com/crossplaneio/crossplane-runtime/pkg/meta"
+	"github.com/crossplaneio/crossplane-runtime/pkg/util"
+	"github.com/crossplaneio/crossplane/apis/stacks/v1alpha1"
+)
+
+const (
+	controllerName = "clusterstackrequest.stacks.crossplane.io"
+
+	reconcileTimeout      = 1 * time.Minute
+	requeueAfterOnSuccess = 10 * time.Second
+
+	packageContentsVolumeName = "package-contents"
+)
+
+var (
+	log              = logging.Logger.WithName(controllerName)
+	resultRequeue    = reconcile.Result{Requeue: true}
+	requeueOnSuccess = reconcile.Result{RequeueAfter: requeueAfterOnSuccess}
+	jobBackoff       = int32(0)
+
+	// podImageNameEnvVar is the env variable for setting the image name used for the stack manager unpack/install process.
+	// When this env variable is not set the parent Pod will be detected and the associated image will be used.
+	// Overriding this variable is only useful when debugging the main stack manager process, since there is no Pod to detect.
+	podImageNameEnvVar = "STACK_MANAGER_IMAGE"
+)
+
+// Reconciler reconciles a Instance object
+type Reconciler struct {
+	sync.Mutex
+	kube       client.Client
+	kubeclient kubernetes.Interface
+	factory
+	executorInfoDiscovery
+	executorInfo *executorInfo
+}
+
+// Controller is responsible for adding the ClusterStackRequest
+// controller and its corresponding reconciler to the manager with any runtime configuration.
+type Controller struct{}
+
+// SetupWithManager creates a new Controller and adds it to the Manager with default RBAC. The Manager will set fields on the Controller
+// and Start it when the Manager is Started.
+func (c *Controller) SetupWithManager(mgr ctrl.Manager) error {
+	r := &Reconciler{
```

Comment:
Should we use the managed reconciler pattern?
https://github.com/crossplaneio/crossplane/blob/master/pkg/controller/aws/cache/managed.go#L64